test/pytest/receiver/test_asapo_transfer.py

```diff
@@ -0,0 +1,163 @@
+    return hidra_metadata
+
+
+@pytest.fixture
+def worker(config):
+    worker_config = config.copy()
+    del worker_config["user_config_path"]
+    worker = AsapoWorker(**worker_config)
+    worker.send_message = create_autospec(worker.send_message)
+    yield worker
+
+
+@pytest.fixture
+def query(config):
+    query = create_autospec(Transfer, instance=True)
+    yield query
+
+
+@pytest.fixture
+def asapo_transfer(worker, query, transfer_config, hidra_metadata):
+    asapo_transfer = AsapoTransfer(worker, query,
+                                   transfer_config['target_host'],
+                                   transfer_config['target_port'],
+                                   transfer_config['target_dir'], transfer_config['reconnect_timeout'])
+    return_list = [[metadata, None] for metadata in hidra_metadata]
+    asapo_transfer.query.get.side_effect = return_list
+    yield asapo_transfer
+
+
+def test_asapo_transfer(caplog, file_list, asapo_transfer):
+
+    x = threading.Thread(target=asapo_transfer.run, args=())
+    x.start()
+    sleep(1)
+    assert f"Send file {file_list[0]}" in caplog.text
+    asapo_transfer.stop()
+    sleep(1)
+    assert "Runner is stopped" in caplog.text
+    x.join()
+
+
+def test_run_transfer(caplog, asapo_transfer):
+
+    x = threading.Thread(target=run_transfer, args=(asapo_transfer, 1))
+    x.start()
+    sleep(2)
+    assert "Retrying connection" in caplog.text
+    asapo_transfer.stop()
+    sleep(1)
+    assert "Runner is stopped" in caplog.text
+    x.join()
+
+
+def test_path_parsing(asapo_transfer, file_list, hidra_metadata):
+    asapo_transfer.stop_run.is_set = Mock()
+    asapo_transfer.stop_run.is_set.side_effect = [False, False, True]
+    asapo_transfer.run()
+
+    asapo_transfer.asapo_worker.send_message.assert_called_with(file_list[0], hidra_metadata[0])
+
+    stream_list = ['foo1/bar1', 'foo1/bar2', 'foo2/bar1', 'foo2/bar2']
+    # ToDo: Refactor: Extract free function or so
+    for i, file_path in enumerate(file_list):
+        data_source, stream, file_idx = asapo_transfer.asapo_worker._parse_file_name(file_path)
+        assert file_idx == 1
+        assert stream == stream_list[i]
+        assert data_source == 'test001'
+
+
+def test_init_query(caplog, asapo_transfer):
+    asapo_transfer.query.initiate.side_effect = KeyError('foo')
+    x = threading.Thread(target=run_transfer, args=(asapo_transfer, 1))
+    x.start()
+    sleep(2)
+    assert "Retrying connection" in caplog.text
+    assert "Stopping query" in caplog.text
+    assert "Starting query" not in caplog.text
+    asapo_transfer.query.stop.assert_called_with()
+    asapo_transfer.stop()
+    sleep(1)
+    assert "Runner is stopped" in caplog.text
+    x.join()
+
+
+def test_start_query(caplog, asapo_transfer):
+    asapo_transfer.query.start.side_effect = KeyError('foo')
+    x = threading.Thread(target=run_transfer, args=(asapo_transfer, 1))
+    x.start()
+    sleep(2)
```

Comment:
Tests would be faster and more stable if you would replace these random sleeps with a `wait_for(predicate, timeout=10)` function call.